Sources/TSCUtility/Triple.swift

```diff
@@ -0,0 +1,211 @@
+///
+public struct Triple: Encodable, Equatable {
+    public let tripleString: String
+
+    public let arch: Arch
+    public let vendor: Vendor
+    public let os: OS
+    public let abi: ABI
+
+    public enum Error: Swift.Error {
+        case badFormat
+        case unknownArch
+        case unknownOS
+    }
+
+    public enum Arch: String, Encodable {
+        case x86_64
+        case x86_64h
+        case i686
+        case powerpc64le
+        case s390x
+        case aarch64
+        case armv7
+        case arm
+        case arm64
+        case arm64e
+        case wasm32
+    }
+
+    public enum Vendor: String, Encodable {
+        case unknown
+        case apple
+    }
+
+    public enum OS: String, Encodable, CaseIterable {
+        case darwin
+        case macOS = "macosx"
+        case linux
+        case windows
+        case wasi
+    }
+
+    public enum ABI: String, Encodable {
+        case unknown
+        case android
+    }
+
+    public init(_ string: String) throws {
+        let components = string.split(separator: "-").map(String.init)
+
+        guard components.count == 3 || components.count == 4 else {
+            throw Error.badFormat
+        }
+
+        guard let arch = Arch(rawValue: components[0]) else {
+            throw Error.unknownArch
+        }
+
+        let vendor = Vendor(rawValue: components[1]) ?? .unknown
+
+        guard let os = Triple.parseOS(components[2]) else {
+            throw Error.unknownOS
+        }
+
+        let abi = components.count > 3 ? Triple.parseABI(components[3]) : nil
+
+        self.tripleString = string
+        self.arch = arch
+        self.vendor = vendor
+        self.os = os
+        self.abi = abi ?? .unknown
+    }
+
+    fileprivate static func parseOS(_ string: String) -> OS? {
+        for candidate in OS.allCases where string.hasPrefix(candidate.rawValue) {
+            return candidate
+        }
+
+        return nil
+    }
+
+    fileprivate static func parseABI(_ string: String) -> ABI? {
+        if string.hasPrefix(ABI.android.rawValue) {
+            return ABI.android
+        }
+        return nil
+    }
+
+    public func isAndroid() -> Bool {
+        return os == .linux && abi == .android
+    }
+
+    public func isDarwin() -> Bool {
+        return vendor == .apple || os == .macOS || os == .darwin
+    }
+
+    public func isLinux() -> Bool {
+        return os == .linux
+    }
+
+    public func isWindows() -> Bool {
+        return os == .windows
+    }
+
+    public func isWASI() -> Bool {
+        return os == .wasi
+    }
+
+    /// Returns the triple string for the given platform version.
+    ///
+    /// This is currently meant for Apple platforms only.
+    public func tripleString(forPlatformVersion version: String) -> String {
+        precondition(isDarwin())
+        return self.tripleString + version
+    }
+
+    public static let macOS = try! Triple("x86_64-apple-macosx")
```

Comment:
This is just being moved over from SwiftPM, right?  Eventually we'll need to fix this.  If it's new code, then we should fix it from the start.

Reply:
Yah, it's just being moved over. Fix it in what sense?

Reply:
I meant to not include the hardcoded architecture.

Reply:
Right, that's only being used for this fallback case anyway: https://github.com/apple/swift-package-manager/blob/master/Sources/SPMBuildCore/Triple.swift#L147-L150

I think we can probably remove that at this point and this static member with it, but I'd like to be conservative for now since this will need to end up in a 5.3.x release.